### ISSUE TITLE
Add a mutable version of Visitor

### ISSUE THREAD

**Issue by canndrew** [dtolnay/syn#203]
I want to be able to walk over an AST and do substitutions/mutations in-place. I think there should be a version of `Visitor` which mutates the AST.

eg.
```rust
pub trait VisitorMut: Sized {
    fn visit_ident(&mut self, _ident: &mut Ident) {}
    fn visit_derive_input(&mut self, derive_input: &mut DeriveInput) {
        walk_derive_input_mut(self, derive_input)
    }
    fn visit_ty(&mut self, ty: &mut Ty) {
        walk_ty_mut(self, ty)
    }
    ... // etc.
}
```

Thoughts?

**Comment by mystor** [comment]
I think that this should probably be part of an effort to unify the visitor-style structs to use a common codepath of some type, as right now they are each managed separately and definitely have bugs in them (as in they miss some subtrees or nodes, or don't have enough visit methods). 

If we could use custom derives it would be awesome to #[derive(Visitors)] on these types.

**Comment by mystor** [comment]
This has been fixed in #206
